spark3/src/main/java/org/apache/iceberg/spark/MigrateAction.java

```diff
@@ -0,0 +1,272 @@
+   */
+  public MigrateAction withAdditionalProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  /**
+   * Adds an additional property to the newly created Iceberg Table. Any properties
+   * with the same key name will be overwritten.
+   * @param key the key of the property to add
+   * @param value the value of the property to add
+   */
+  public MigrateAction withAdditionalProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  public MigrateAction(SparkSession spark, Identifier destTableName, Identifier sourceTableName) {
+    this.spark = spark;
+    this.sourceTableName = sourceTableName;
+    this.destTableName = destTableName;
+    this.destCatalog = getSparkSessionCatalogOrFail(spark, destTableName);
+
+    try {
+      sourcePartitionSpec = SparkSchemaUtil.specForTable(spark, Spark3Util.toTableIdentifier(sourceTableName));
+    } catch (AnalysisException e) {
+      throw new IllegalArgumentException("Cannot determining partitioning of " + sourceTableName.toString(), e);
+    }
+
+    try {
+      this.sourceTable =
+        spark.sessionState().catalog().getTableMetadata(Spark3Util.toTableIdentifier(sourceTableName));
+    } catch (NoSuchTableException | NoSuchDatabaseException e) {
+      throw new IllegalArgumentException(String.format("Could not find source table %s", sourceTableName), e);
+    }
+    validateSourceTable(sourceTable, ALLOWED_SOURCES);
+
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+    this.sessionCatalogReplacement = !sourceTableProvider.equals("iceberg") && sourceTableName.equals(destTableName);
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceTable.storage().locationUri().get());
+    this.destDataLocation = sourceTableLocation;
+    this.destMetadataLocation = sourceTableLocation + "/" + ICEBERG_METADATA_FOLDER;
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable, Set<String> supportedSourceTableProviders) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+
+    if (!supportedSourceTableProviders.contains(sourceTableProvider)) {
+      throw new IllegalArgumentException(
+          String.format("Cannot create an Iceberg table from source provider: %s", sourceTableProvider));
+    }
+    if (sourceTable.storage().locationUri().isEmpty()) {
+      throw new IllegalArgumentException("Cannot create an Iceberg table from a source without an explicit location");
+    }
+  }
+
+  private static Map<String, String> extraIcebergTableProps(String tableLocation, String metadataLocation) {
+    return ImmutableMap.of(
+        TableProperties.WRITE_METADATA_LOCATION, metadataLocation,
+        TableProperties.WRITE_NEW_DATA_LOCATION, tableLocation,
+        "migrated", "true");
+  }
+
+  private static SparkSessionCatalog getSparkSessionCatalogOrFail(
```

Comment:
Style nit: Both "Spark" and "OrFail" are implied, so I think they are just making this method name longer. It could be `getSessionCatalog`.